File: code/java/collections/06_Fold.java
private &ltT&gt Double calculateTotals(final Grid&ltT&gt grid) {
    return grid.getRows()
        .stream()
        .flatMap(e -> e.getColumns().stream())
        .collect(Collectors.groupingBy(Column::getTitle))
        .entrySet()
        .stream()
        .flatMap(e -> e.getValue().stream())
        .map(Column::getValue)
        .reduce(
            0.0,
            (accumulator, value) ->
                value instanceof Number
                    ? accumulator + ((Number) (value)).doubleValue()
                    : accumulator,
            Double::sum
        );

@Data // Lombok
class Grid&ltT&gt {
    private final List<Row&ltT&gt> rows;
}

@Data // Lombok
class Row&ltT&gt {
    private final List<Column&ltT&gt> columns;
}

@Data // Lombok
class Column&ltT&gt {
    private final String title;
    private final T value;
}

File: code/kotlin/collections/06_Fold.kt
fun &ltT&gt calculateTotals(grid: Grid&ltT&gt) =
    grid.rows
        .flatMap(Row&ltT&gt::columns)
        .groupingBy(Column&ltT&gt::title)
        .fold(0.0) {
            accumulator,
            (_, value) ->
                accumulator +
                        when (value) {
                            is Number -> value.toDouble()
                            else -> 0.0
                        }
        }

data class Grid&ltT&gt(val rows: List<Row&ltT&gt>)
data class Row&ltT&gt(val columns: List<Column&ltT&gt>)
data class Column&ltT&gt(val title: String, val value: T)

// https://kotlinexpertise.com/kotlin-productivity/